### DISCUSSION TITLE
On transaction level simulation.

### DISCUSSION THREAD

**Discussion by Readon** [SpinalHDL/SpinalHDL#943] [Ideas]
I have talked with people that complain about the lacking of transaction level simulation of SpinalHDL.
Is there any ideas that about transaction level API for simulation?

I think the agent of [AxiLite4Driver](https://github.com/SpinalHDL/SpinalHDL/blob/018dcab750db5a2cb4637f041b6e598466534449/lib/src/main/scala/spinal/lib/bus/amba4/axilite/sim/AxiLite4Driver.scala#L7) can be treated as a transaction level driver, which have read and write functions.
Although it's not in non-blocking style but blocking, is API what people want for transaction level agent?

**Comment by numero-744** [comment]
Something like this?

```scala
    val write = dut.master.Write(A, Data(A))
    write.run()
    val read = dut.master.Read(A)
    read.run()
    assert(read.result == Data(A))
    assert(read.duration == 2)
    assert(read.waitStates == 0)
```

https://github.com/SpinalHDL/SpinalHDL/pull/879

**Comment by kleinai** [comment]
I'd like to see something like what VUnit does where you can queue up read and write commands. I don't believe that's possible with the Axi4 driver as is. It would make testing memory mapped components easier.

**Reply by numero-744** [reply]
And this?

```scala
    val write = dut.master.Write(A, Data(A))
    val read = dut.master.Read(A)
    val sequence = write >> read
    sequence.run()
    assert(read.result == Data(A))
    assert(read.duration == 2)
    assert(read.waitStates == 0)
```

#879

**Comment by saahm** [comment]
I'm a bit confused on how something utilizing pin-accurate/cycle-accurate modeling can be considered transaction level?

The term "Transaction Level" can have a very different meaning when looking it from the point of view (namely people who do Transaction Level Modeling for example, so Electronic System Level Design).
I could give you a few more sources but lets go with the least strong one here: the Wiki article on ESL (https://en.wikipedia.org/wiki/Electronic_system-level_design_and_verification) gives a summary on for example what TLM means in its context. `[...] TLM, where operations of one component on another are modelled with a simple method call between the objects modelling each component. This abstraction gives a considerable speed up over cycle-accurate modelling, since thousands of net-level events in the real system can be represented by simply passing a pointer, e.g. to model that an Ethernet packet has been received, SystemC is often used.`

In my background I've worked with Virtual Prototyping and SystemC TLM already thus when I read "Transaction Level" here I got a bit confused as SpinalHDL only touches that 'borderline' so to say (its still HDL thus 'wanting' cycle-accurate/pin-accurate modeling).

**Reply by Readon** [reply]
Yes, the transaction level modeling in SystemC is a little bit different, what their target is to simulate the time events of digital system.
I think for SpinalHDL the transaction level simulation is much like @kleinai described above.
